tests/kernel/smp_suspend/src/main.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+#include <zephyr/kernel.h>
+
+#if CONFIG_MP_MAX_NUM_CPUS < 2
+#error "SMP test requires at least two CPUs!"
+#endif
+
+#define STACK_SIZE 1024
+
+#define NUM_THREADS 6
+
+K_THREAD_STACK_ARRAY_DEFINE(thread_stack, NUM_THREADS, STACK_SIZE);
+struct k_thread thread[NUM_THREADS];
+
+volatile uint64_t thread_counter[NUM_THREADS];
+
+static void thread_entry(void *p1, void *p2, void *p3)
+{
+	struct k_thread *resume_thread = p1;
+	unsigned int self_index = (unsigned int)(uintptr_t)p2;
+
+	while (1) {
+		if (resume_thread != NULL) {
+			k_thread_resume(resume_thread);
+		}
+
+		thread_counter[self_index]++;
+
+		if (self_index != 0) {
+			k_thread_suspend(k_current_get());
+		}
+	}
+}
+
+ZTEST(smp_suspend_resume, test_smp_thread_suspend_resume_stress)
+{
+	unsigned int  i;
+	uint64_t counter[NUM_THREADS] = {};
+
+	/* Create the threads */
+
+	printk("Starting ...\n");
+
+	for (i = 0; i < NUM_THREADS; i++) {
+
+		k_thread_create(&thread[i], thread_stack[i],
+				STACK_SIZE, thread_entry,
+				i < (NUM_THREADS - 1) ? &thread[i + 1] : NULL,
+				(void *)(uintptr_t)i, NULL,
+				10 - i, 0, K_FOREVER);
+
+		k_thread_suspend(&thread[i]);
+
+		k_thread_start(&thread[i]);
```

Comment:
Might be worth a comment explaining here that you want the thread "launched" in a suspended state so it doesn't run until k_thread_resume() is called.

That said: that's a good edge case to test but it probably wouldn't hurt to be exercising an async k_thread_suspend() too, given that this is a stress test.